cirq/linalg/transformations.py

```diff
@@ -304,3 +317,152 @@ def partial_trace(tensor: np.ndarray,
     left_indices = [keep_map[i] if i in keep_set else i for i in range(ndim)]
     right_indices = [ndim + i if i in keep_set else i for i in left_indices]
     return np.einsum(tensor, left_indices + right_indices)
+
+
+def wavefunction_partial_trace(
+        wavefunction: np.ndarray,
+        keep_indices: List[int],
+        *,
+        atol: Union[int, float] = 1e-8) -> Tuple[Tuple[float, np.ndarray]]:
+    """Attempts to factor a wavefunction into two parts and return one of them.
+
+    The input wavefunction must have shape `(2, 2, ..., 2)` or `(2 ** N)` where
+    `wavefunction` is expressed over N qubits. Attempts to factor the
```

Comment:
The optimization for the pure case should only be mentioned in passing at the end, because it's not something the user needs to understand in order to use the method.